Tests/NIOTests/ChannelTests.swift

```diff
@@ -2019,4 +2019,86 @@ public class ChannelTests: XCTestCase {
         try allDone.futureResult.wait()
         XCTAssertNoThrow(try sc.syncCloseAcceptingAlreadyClosed())
     }
+
+    func testSocketFailingAsyncCorrectlyTearsTheChannelDownAndDoesntCrash() throws {
+        // regression test for #302
+        enum DummyError: Error { case dummy }
+        class SocketFailingAsyncConnect: Socket {
+            private var firstReadHappened = false
+            init() throws {
+                try super.init(protocolFamily: PF_INET, type: Posix.SOCK_STREAM, setNonBlocking: true)
+            }
+            override func connect(to address: SocketAddress) throws -> Bool {
+                _ = try super.connect(to: address)
+                return false
+            }
+            override func finishConnect() throws {
+                throw DummyError.dummy
+            }
+        }
+        class VerifyThingsAreRightHandler: ChannelInboundHandler {
+            typealias InboundIn = Never
+            private let allDone: EventLoopPromise<Void>
+            enum State {
+                case fresh
+                case registered
+                case unregistered
+            }
+            private var state: State = .fresh
+
+            init(allDone: EventLoopPromise<Void>) {
+                self.allDone = allDone
+            }
+            deinit {
+                XCTAssertEqual(.unregistered, self.state)
+            }
+            func channelActive(ctx: ChannelHandlerContext) {
+                XCTFail("should never become active")
+            }
+            func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+                XCTFail("should never read")
+            }
+            func channelReadComplete(ctx: ChannelHandlerContext) {
+                XCTFail("should never readComplete")
+            }
+            func channelRegistered(ctx: ChannelHandlerContext) {
+                XCTAssertEqual(.fresh, self.state)
+                self.state = .registered
+            }
+            func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+                XCTFail("pipeline shouldn't be told about connect error")
+            }
+            func channelUnregistered(ctx: ChannelHandlerContext) {
+                XCTAssertEqual(.registered, self.state)
+                self.state = .unregistered
+                self.allDone.succeed(result: ())
+            }
+        }
+        let group = MultiThreadedEventLoopGroup(numThreads: 2)
+        defer {
+            XCTAssertNoThrow(try group.syncShutdownGracefully())
+        }
+        let serverEL = group.next()
+        let clientEL = group.next()
+        precondition(serverEL !== clientEL)
```

Comment:
I don't think we need the two threads here: it's totally fine for them to run on the one thread.

Reply:
totally unpicking my copy&paste here :D